### PR TITLE
REG: read_excel with engine specified raises on non-path/non-buffer

### LINKED ISSUE #39528: BUG: read_excel with Workbook and engine="openpyxl" raises ValueError
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample

```python
from openpyxl import load_workbook
from pandas import read_excel

wb = load_workbook("testfile.xlsx")
read_excel(wb, engine="openpyxl")
```

#### Problem description

In pandas 1.1.5, the above code completes with no problems.

In pandas 1.2.1, it causes the following exception:

```
Traceback (most recent call last):
  File "c:/Users/akaijanaho/scratch/pandas-openpyxl-bug/bug.py", line 5, in <module>
    read_excel(wb, engine="openpyxl")
  File "C:\Users\akaijanaho\scratch\pandas-openpyxl-bug\venv\lib\site-packages\pandas\util\_decorators.py", line 299, in wrapper
    return func(*args, **kwargs)
  File "C:\Users\akaijanaho\scratch\pandas-openpyxl-bug\venv\lib\site-packages\pandas\io\excel\_base.py", line 336, in read_excel
    io = ExcelFile(io, storage_options=storage_options, engine=engine)
  File "C:\Users\akaijanaho\scratch\pandas-openpyxl-bug\venv\lib\site-packages\pandas\io\excel\_base.py", line 1057, in __init__
    ext = inspect_excel_format(
  File "C:\Users\akaijanaho\scratch\pandas-openpyxl-bug\venv\lib\site-packages\pandas\io\excel\_base.py", line 938, in inspect_excel_format
    with get_handle(
  File "C:\Users\akaijanaho\scratch\pandas-openpyxl-bug\venv\lib\site-packages\pandas\io\common.py", line 558, in get_handle
    ioargs = _get_filepath_or_buffer(
  File "C:\Users\akaijanaho\scratch\pandas-openpyxl-bug\venv\lib\site-packages\pandas\io\common.py", line 371, in _get_filepath_or_buffer
    raise ValueError(msg)
ValueError: Invalid file path or buffer object type: <class 'openpyxl.workbook.workbook.Workbook'>
```

The documentation does not specify Workbook as an acceptable value type for io, but supporting it seems reasonable and accords with the 1.1.5 behavior.

In my use case, we mainly parse an Excel file with openpyxl but use pandas with a specific sub-problem. We would like to reuse the same Workbook instead of having pandas re-read the file.
</details>

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #39528
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — APPROVED** [review]
@twoertwein ?

File: pandas/io/excel/_base.py [hunk 1]
```diff
@@ -1069,26 +1069,37 @@ def __init__(
...
-            ext = "xls"
-        else:
-            ext = inspect_excel_format(
-                content_or_path=path_or_buffer, storage_options=storage_options
-            )
-
+        ext = None
         if engine is None:
+            # Only determine ext if it is needed
+            if xlrd_version is not None and isinstance(path_or_buffer, xlrd.Book):
```
Comment (twoertwein) [inline pandas/io/excel/_base.py:1075]: could probably remove the nesting (I assume we always have to use `ext="xls"` even if `engine` is not `None`):

```py
if xlrd_version is not None and isinstance(path_or_buffer, xlrd.Book):
    ...
elif engine is not None:
    ....
```

**Comment by twoertwein** [comment]
looks good to me!

File: pandas/tests/io/excel/test_openpyxl.py [hunk 2]
```diff
@@ -116,3 +116,12 @@ def test_to_excel_with_openpyxl_engine(ext):
...
 
         styled.to_excel(filename, engine="openpyxl")
+
+
+def test_read_workbook(datapath, ext):
+    # GH 39528
+    filename = datapath("io", "data", "excel", "test1" + ext)
+    wb = openpyxl.load_workbook(filename)
+    result = pd.read_excel(wb, engine="openpyxl")
+    expected = pd.read_excel(filename)
```
Comment (twoertwein) [inline pandas/tests/io/excel/test_openpyxl.py:126]: It cannot hurt to call `wb.close()` before `tm.assert_frame_equal`

Comment (rhshadrach) [reply pandas/tests/io/excel/test_openpyxl.py:127]: Thanks @twoertwein, added.

**Comment by jreback** [comment]
if you can merge master and ping on greem

**Comment by rhshadrach** [comment]
This PR enables the possibility that the sheet in `get_sheet_data` is not read-only. We only want to apply the fix from #39486 in the case the sheet is read-only (which is always the case when reading from a path or buffer). Unfortunately it seems to me that openpyxl provides no way of telling if a sheet is read-only or not; for this I've opened [an issue upstream](https://foss.heptapod.net/openpyxl/openpyxl/-/issues/1605).

The method `reset_dimensions` only exists on read-only sheets, and so it can be used for this purpose (even if not ideal).

I've added tests for this and to ensure the fix from #39486 works on non-read-only workbooks.

**Comment by rhshadrach** [comment]
@jreback - green

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rhshadrach

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x

**Comment by jreback** [comment]
cc @simonjayhawkins

**Comment by simonjayhawkins** [comment]
@rhshadrach There have been some changes to master since 1.2 that were not backported (#38710 and  #39008) and trying to apply these changes, I'm getting

```
=================================================== short test summary info ===================================================
FAILED pandas/tests/io/excel/test_readers.py::TestReaders::test_corrupt_bytes_raises[engine_and_read_ext0] - AttributeError:...
FAILED pandas/tests/io/excel/test_readers.py::TestExcelFileRead::test_excel_read_binary[engine_and_read_ext0] - AttributeErr...
================================== 2 failed, 1398 passed, 146 skipped, 10 xfailed in 25.65s ===================================
```

**Comment by simonjayhawkins** [comment]
ahh , needed a `content=`

**Pull request merged**